nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java

```diff
@@ -172,6 +180,8 @@ public static Object convertType(final Object value, final DataType dataType, fi
 
                 return convertType(value, chosenDataType, fieldName, charset);
             }
+            case DECIMAL:
+                return toDecimal(value, fieldName).setScale(((DecimalDataType)dataType).getScale(), RoundingMode.HALF_UP);
```

Comment:
Why not pass the scale (and ideally precision) to the `toDecimal` method and set the scale and rounding mode (and check precision) within that method?